plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,254 @@
+{-# LANGUAGE ConstraintKinds    #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveLift         #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE OverloadedLists    #-}
+{-# LANGUAGE PatternSynonyms    #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE ViewPatterns       #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:profile-all #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:conservative-optimisation #-}
+
+module PlutusTx.DataMap (
+  Map,
+  singleton,
+  empty,
+  keys,
+  lookup,
+  member,
+  insert,
+  delete,
+  union,
+  unionWith,
+  filter,
+  mapMaybe,
+  all,
+  mapThese,
+  map,
+  foldr,
+  revAppend,
+  allPair,
+  null,
+  foldMapPair,
+  foldMap,
+  pattern MNil,
+  pattern MCons,
+) where
+
+import Control.DeepSeq (NFData)
+import Data.Data (Data)
+import GHC.Generics (Generic)
+import Language.Haskell.TH.Syntax as TH (Lift)
+import PlutusTx.AsData qualified as AsData
+import PlutusTx.Builtins qualified as B
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.DataList (List (List_), pattern Cons, pattern Nil)
+import PlutusTx.DataList qualified as List
+import PlutusTx.DataPair (DataElem, Pair, fst, pattern Pair, snd)
+import PlutusTx.DataPair qualified as Pair
+import PlutusTx.DataThese (These, pattern That, pattern These, pattern This)
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (all, any, concat, filter, foldMap, foldr, fst, map, null, revAppend,
+                         snd, toList)
+import Prelude qualified as H
+
+newtype Map k v = Map_ BuiltinData
+  deriving stock (Generic, Data, H.Show)
+  deriving anyclass (NFData)
+  deriving newtype (P.ToData, P.FromData, P.UnsafeFromData)
+
+pattern MNil :: Map k v
+pattern MNil <- (Map_ (BI.unsafeDataAsMap -> B.headMaybe -> Nothing))
+  where
+    MNil = Map_ . BI.mkMap . BI.mkNilPairData $ BI.unitval
+
+pattern MCons :: (DataElem k, DataElem v) => Pair k v -> Map k v -> Map k v
+pattern MCons x xs <-
+  (Map_
+    (BI.unsafeDataAsMap ->
+      B.unsafeUncons ->
+        (Pair.fromBuiltinPair -> x, BI.mkMap -> Map_ -> xs)
+    )
+  )
+  where
+    MCons a (Map_ as) =
+      let aPair = Pair.toBuiltinPair a
+          builtinList = BI.unsafeDataAsMap as
+       in Map_ $ BI.mkMap $ BI.mkCons aPair builtinList
+
+{-# COMPLETE MNil, MCons #-}
+
+-- I think ^ doesn't work because the Data encoding doesn't know that
+-- BuiltinList (BuiltinPair BuiltinData BuiltinData) is an instance of BuiltinList BuiltinData
+-- also Map is a completely different constructor from List ( this is the actual reason)
+
+instance (DataElem k, DataElem v, Eq k, Semigroup v) =>  Semigroup (Map k v) where
+  (<>) = unionWith (<>)
+
+instance (DataElem k, DataElem v, Eq k, Semigroup v) => Monoid (Map k v) where
+  mempty = empty
+
+empty :: Map k v
+empty = MNil
+
+lookup :: (DataElem k, DataElem v, Eq k) => k -> Map k v -> Maybe v
```

Comment:
All these `DataElem` constraints could also lead to higher costs, since sometimes only one of `ToData` and `UnsafeFromData` is required.